File: jsh/lib/mathx/mat/matrix.js
'use strict';

const _mat = require('@jsh/mathx/mat');

function isNumber(value) {
    return typeof value === 'number';
}

function isInteger(value) {
    return Number.isInteger(value);
}

function validateDimension(name, value, label) {
    if (!isInteger(value)) {
        throw new Error(`${name}: ${label} should be an integer`);
    }
    if (value < 0) {
        throw new Error(`${name}: ${label} should be non-negative`);
    }
}

function validateNumericArray(name, data, expectedLength) {
    if (data === undefined || data === null) {
        return;
    }
    if (!Array.isArray(data)) {
        throw new Error(`${name}: data should be an array`);
    }
    for (const value of data) {
        if (!isNumber(value)) {
            throw new Error(`${name}: data should contain only numbers`);
        }
    }
    if (expectedLength !== undefined && data.length !== expectedLength) {
        throw new Error(`${name}: data length should be ${expectedLength}`);
    }
}

function Dense(rows, cols, data) {
    if (arguments.length > 0) {
        validateDimension('Dense', rows, 'rows');
    }
    if (arguments.length > 1) {
        validateDimension('Dense', cols, 'cols');
    }
    if (arguments.length > 2) {
        validateNumericArray('Dense', data, rows * cols);
    }
    if (arguments.length === 2) {
        return _mat.Dense(rows, cols, null);
    } else if (arguments.length === 3) {
        return _mat.Dense(rows, cols, data);
    } else {
        return _mat.Dense(0, 0, null);
    }
}

function SymDense(n, data) {
    if (arguments.length > 0) {
        validateDimension('SymDense', n, 'size');
    }
    if (arguments.length > 1) {
        validateNumericArray('SymDense', data, n * n);
    }
    return _mat.SymDense(n, data);
}

function VecDense(n, data) {
    if (arguments.length > 0) {
        validateDimension('VecDense', n, 'size');
    }
    if (arguments.length > 1) {
        validateNumericArray('VecDense', data, n);
    }
    return _mat.VecDense(n, data);
}

function QR() {
    return _mat.QR(...arguments);
}

function format(matrix, opts) {
    if (arguments.length === 0) {
        return undefined;
    }
    // if (!matrix || typeof matrix !== 'object' || !('$' in matrix)) {
    //     return undefined;
    // }

    if (arguments.length > 1 && opts !== undefined && opts !== null) {
        if (typeof opts !== 'object' || Array.isArray(opts)) {
            throw new Error('format: invalid options');
        }
        if (opts.format !== undefined && typeof opts.format !== 'string') {
            throw new Error('format: invalid options');
        }
        if (opts.prefix !== undefined && typeof opts.prefix !== 'string') {
            throw new Error('format: invalid options');
        }
        if (opts.excerpt !== undefined && !isInteger(opts.excerpt)) {
            throw new Error('format: invalid options');
        }
        if (opts.squeeze !== undefined && typeof opts.squeeze !== 'boolean') {
            throw new Error('format: invalid options');
        }
    }
    return _mat.format(matrix, opts);
}

module.exports = {
    Dense,
    format,
    QR,
    SymDense,
    VecDense,
};

File: jsh/lib/mathx/mat/matrix.go
package mat

import (
	"context"
	_ "embed"
	"fmt"

	"github.com/dop251/goja"
	"gonum.org/v1/gonum/mat"
)

//go:embed matrix.js
var matrix_js []byte

func Files() map[string][]byte {
	return map[string][]byte{
		"mathx/mat.js": matrix_js,
	}
}

func Module(_ context.Context, rt *goja.Runtime, module *goja.Object) {
	// m = require("@jsh/mathx/mat")
	o := module.Get("exports").(*goja.Object)
	// format("%v", m, opts...)
	o.Set("format", Format)
	// new Dense(rows, cols, []float64)
	o.Set("Dense", new_dense)
	// new SymDense(dims, []float64)
	o.Set("SymDense", new_symDense)
	// new QR()
	o.Set("QR", new_qr)
	// new VecDense(n, []float64)
	o.Set("VecDense", new_vecDense)
}

func new_dense(rows, cols int, data []float64) *mat.Dense {
	if rows <= 0 || cols <= 0 {
		return &mat.Dense{}
	}
	return mat.NewDense(rows, cols, data)
}

func new_symDense(n int, data []float64) *mat.SymDense {
	if n <= 0 {
		return &mat.SymDense{}
	}
	return mat.NewSymDense(n, data)
}

func new_vecDense(n int, data []float64) *mat.VecDense {
	if n <= 0 {
		return &mat.VecDense{}
	}
	return mat.NewVecDense(n, data)
}

func new_qr() *mat.QR {
	return &mat.QR{}
}

type FormatOption struct {
	Format  string `json:"format"`
	Prefix  string `json:"prefix,omitempty"`
	Excerpt int    `json:"excerpt,omitempty"`
	Squeeze bool   `json:"squeeze,omitempty"`
}

func Format(v mat.Matrix, opts FormatOption) string {
	if opts.Format == "" {
		opts.Format = "%v"
	}

	o := []mat.FormatOption{}
	if opts.Prefix != "" {
		o = append(o, mat.Prefix(opts.Prefix))
	}
	if opts.Excerpt > 0 {
		o = append(o, mat.Excerpt(opts.Excerpt))
	}
	if opts.Squeeze {
		o = append(o, mat.Squeeze())
	}
	f := mat.Formatted(v, o...)
	return fmt.Sprintf(opts.Format, f)
}
